plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1167 @@
+
+                                    lastMsgOffset = messageObjPtr->offset();
+                                    ++consumedRecCount;
+
+                                    // Give opportunity for consumer to pull in any additional messages
+                                    consumerPtr->handle()->poll(0);
+                                }
+                                break;
+
+                            case RdKafka::ERR__TIMED_OUT:
+                                // No new messages arrived and we timed out waiting
+                                ++attemptNum;
+                                consumerPtr->handle()->poll(timeoutWait);
+                                break;
+
+                            case RdKafka::ERR__PARTITION_EOF:
+                                // We reached the end of the messages in the partition
+                                #ifdef _DEBUG
+                                    DBGLOG("Kafka: EOF reading message from partition %d", messageObjPtr->partition());
+                                #endif
+                                shouldRead = false;
+                                break;
+
+                            case RdKafka::ERR__UNKNOWN_PARTITION:
+                                // Unknown partition; don't throw an error here because
+                                // in some configurations (e.g. more Thor slaves than
+                                // partitions) not all consumers will have a partition
+                                // to read
+                                #ifdef _DEBUG
+                                    DBGLOG("Kafka: Unknown partition while trying to read");
+                                #endif
+                                shouldRead = false;
+                                break;
+
+                            case RdKafka::ERR__UNKNOWN_TOPIC:
+                                throw MakeStringException(-1, "Kafka: Error while reading message: '%s'", messageObjPtr->errstr().c_str());
+                                break;
+                        }
+                    }
+                    catch (...)
+                    {
+                        delete(messageObjPtr);
+                        throw;
+                    }
+
+                    delete(messageObjPtr);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    void KafkaStreamedDataset::stop()
+    {
+        shouldRead = false;
+    }
+
+    //--------------------------------------------------------------------------
+
+    Poller::Poller(KafkaObj* _parentPtr, __int32 _pollTimeout)
+        :   Thread("Kafka::Poller"),
+            parentPtr(_parentPtr),
+            pollTimeout(_pollTimeout)
+    {
+        shouldRun = false;
+        running = false;
+    }
+
+    Poller::~Poller()
+    {
+
+    }
+
+    void Poller::start()
+    {
+        if (!running)
+        {
+            shouldRun = true;
```

Comment:
There's a race condition if stop() is called just after start(), before the started thread has set running=true